components/layout_2020/query.rs

```diff
@@ -525,8 +529,238 @@ fn is_eligible_parent(fragment: &BoxFragment) -> bool {
+                let is_preformatted_element =
+                    node.type_id() == LayoutNodeType::Element(LayoutElementType::HTMLPreElement);
+
+                let is_final_character_whitespace = text_content
+                    .chars()
+                    .last()
+                    .filter(|c| c.is_ascii_whitespace())
+                    .is_some();
+
+                let white_space_collapse = style.clone_white_space_collapse();
+
+                // To account for the deviations from normal white-space processing, omit the final
+                // character of the text if it meets specific conditions
+                let num_chars_to_take = match (
+                    white_space_collapse == WhiteSpaceCollapseValue::Preserve,
+                    is_preformatted_element,
+                    is_final_character_whitespace,
+                    iterator.peek(),
+                ) {
+                    // Remove trailing whitespace before <br> elements
+                    (false, false, true, Some(el))
+                        if el.type_id() ==
+                            LayoutNodeType::Element(LayoutElementType::HTMLBRElement) =>
+                    {
+                        text_content.len() - 1
+                    },
+                    // Remove trailing whitespace if this is the last line of the block
+                    (false, false, true, None) => text_content.len() - 1,
+                    _ => text_content.len(),
+                };
+
+                let with_white_space_rules_applied = WhitespaceCollapse::new(
+                    text_content.chars().take(num_chars_to_take),
+                    white_space_collapse,
+                    white_space_collapse != WhiteSpaceCollapseValue::Preserve,
+                );
+
+                let result = TextTransformation::new(
+                    with_white_space_rules_applied,
+                    style.clone_text_transform().case(),
+                )
+                .collect::<String>();
+
+                results.push(InnerOrOuterTextItem::Text(result));
+
+                continue;
+            },
+            LayoutNodeType::Element(LayoutElementType::HTMLBRElement) => {
+                // Step 5: If node is a br element, then append a string containing a single U+000A
+                // LF code point to items.
+                results.push(InnerOrOuterTextItem::Text(String::from("\u{000A}")));
+            },
+            LayoutNodeType::Element(LayoutElementType::HTMLParagraphElement) => {
+                // Step 8: If node is a p element, then append 2 (a required line break count) at
+                // the beginning and end of items.
+                results.insert(0, InnerOrOuterTextItem::RequiredLineBreakCount(2));
+                results.push(InnerOrOuterTextItem::RequiredLineBreakCount(2));
+            },
+            _ => {},
+        }
+
+        match display {
+            Display::TableCell if iterator.peek().is_some() => {
```

Comment:
the condition `if iterator.peek().is_some()` makes no sense here, we need a fn which determines whether it's the last cell in the row.

Reply:
Maybe we can invert the logic here, instead of emitting the TAB/LF characters after each Cell/Row we instead emit one before the contents if it's not the first Cell/Row in a Table/Row? That way we should get the same results and don't have to peek. 

Might give different results with strange nesting, though I'm not sure what the standards say about the required behaviour there.

Reply:
I agree, when you read the spec, it's as though the have approached things starting at a text node and moving outwards to the parents, but we approach them from parent to child.

> we instead emit one before the contents if it's not the first Cell/Row in a Table/Row?

This makes sense to me - In the specific case of cells, the logic that seems sensible in my mind:

```
if is_a_text_node() {
   append_text()
   
   if parent_node_is_display_table_cell && _not_the_last_display_table_cell_in_its_row() {
       append_tab_char()
   }
}
```

I've just pushed a very questionable implementation of this, it does get us more passes but we would have to find a way to deal with table rows.